compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala

```diff
@@ -91,57 +85,19 @@ class ExtractAPI extends Phase {
 
-  // Why we only write to early output in the first run?
-  // ===================================================
-  // TL;DR the point of pipeline compilation is to start downstream projects early,
-  // so we don't want to wait for suspended units to be compiled.
-  //
-  // But why is it safe to ignore suspended units?
-  // If this project contains a transparent macro that is called in the same project,
-  // the compilation unit of that call will be suspended (if the macro implementation
-  // is also in this project), causing a second run.
-  // However before we do that run, we will have already requested sbt to begin
-  // early downstream compilation. This means that the suspended definitions will not
-  // be visible in *early* downstream compilation.
-  //
-  // However, sbt will by default prevent downstream compilation happening in this scenario,
-  // due to the existence of macro definitions. So we are protected from failure if user tries
-  // to use the suspended definitions.
-  //
-  // Additionally, it is recommended for the user to move macro implementations to another project
-  // if they want to force early output. In this scenario the suspensions will no longer occur, so now
-  // they will become visible in the early-output.
-  //
-  // See `sbt-test/pipelining/pipelining-scala-macro` and `sbt-test/pipelining/pipelining-scala-macro-force`
-  // for examples of this in action.
-  //
-  // Therefore we only need to write to early output in the first run. We also provide the option
-  // to diagnose suspensions with the `-Yno-suspended-units` flag.
-  private def writeSigFiles(units: List[CompilationUnit], writer: Pickler.EarlyFileWriter)(using Context): Unit = {
-    try
-      for
-        unit <- units
-        (cls, pickled) <- unit.pickled
-        if cls.isDefinedInCurrentRun
-      do
-        val internalName =
-          if cls.is(Module) then cls.binaryClassName.stripSuffix(str.MODULE_SUFFIX).nn
-          else cls.binaryClassName
-        val _ = writer.writeTasty(internalName, pickled())
-    finally
-      writer.close()
-      if ctx.settings.verbose.value then
-        report.echo("[sig files written]")
-    end try
-  }
-
   private def recordNonLocalClasses(nonLocalClassSymbols: mutable.HashSet[Symbol], cb: interfaces.IncrementalCallback)(using Context): Unit =
     for cls <- nonLocalClassSymbols do
       val sourceFile = cls.source
       if sourceFile.exists && cls.isDefinedInCurrentRun then
         recordNonLocalClass(cls, sourceFile, cb)
-    cb.apiPhaseCompleted()
-    cb.dependencyPhaseCompleted()
+    for holder <- ctx.asyncTastyPromise do
+      import scala.concurrent.ExecutionContext.Implicits.global
+      // do not expect to be completed with failure
+      holder.promise.future.foreach: state =>
+        if !state.hasErrors then
+          // We also await the promise at GenBCode to emit warnings/errors
+          cb.apiPhaseCompleted()
+          cb.dependencyPhaseCompleted()
```

Comment:
Does zinc/sbt depend on these callbacks being called before the overall compilation finishes? If yes, we have a theoretical race between `GenBCode` completing and these two callbacks completing. I don't know if it could happen in practice, as I guess these callbacks should be much quicker than `GenBCode`, but it's still worth paying attention to.

Ideally, we would make this a `.map` on the initial promise, and `GenBCode` would wait on the result of that `map`. This would ensure that we don't have any race, without slowing down the expected scenario where callbacks complete well before `GenBCode` gets to run. It would also make it clearer that asynchronous errors are not dropped; they are forwarded through `map` to next handler in line.

Reply:
new commit composes this signal with the tasty written signal before blocking at the backend.